src/nfa/noodlify.cc

```diff
@@ -38,78 +39,146 @@ size_t get_num_of_permutations(const SegNfa::Segmentation::EpsilonDepthTransitio
     return num_of_permutations;
 }
 
-/**
- * Create noodles for computed epsilon depths.
- * @param[in] aut Segment automaton to create noodles for.
- * @param[in] include_empty Whether to also include empty noodles.
- * @param[in] epsilon_depths Computed list of epsilon transitions for each depth.
- * @return A Sequence of noodles (noodle automata).
- */
-AutSequence create_noodles(const SegNfa::SegNfa& aut, bool include_empty,
-                           const SegNfa::Segmentation::EpsilonDepthTransitions& epsilon_depths)
-{
-    // Compute number of all combinations of ε-transitions with one ε-transitions from each depth.
-    size_t num_of_permutations{ get_num_of_permutations(epsilon_depths) };
-    int epsilon_depths_size{ static_cast<int>(epsilon_depths.size()) };
-    int maximal_depth{ epsilon_depths_size - 1 };
-    AutSequence noodles{}; // Store the generated noodles.
-    Nfa noodle{}; // Noodle created for each combination of epsilon transitions.
-    // Indices of a transition in each depth to compute the noodle for.
-    std::vector<size_t> transition_indices(epsilon_depths_size, 0);
-    // For each combination of ε-transitions, create the automaton.
-    for (size_t index{ 0 }; index < num_of_permutations; ++index)
-    {
-        for (int depth{ maximal_depth }; depth >= 0; --depth)
-        {
-            size_t computed_index{ index };
-            for (int previous_depth{ 0 }; previous_depth < depth; ++previous_depth)
-            {
-                computed_index /= epsilon_depths.at(previous_depth).size();
-            }
+} // namespace
 
-            transition_indices.at(depth) = computed_index % epsilon_depths.at(depth).size();
+SegNfa::NoodleSequence SegNfa::noodlify(const SegNfa& aut, const Symbol epsilon, bool include_empty) {
+    Segmentation segmentation{ aut, epsilon };
+    const auto& segments{ segmentation.get_segments_raw() };
+
+    if (segments.size() == 1) {
+        std::shared_ptr<Nfa> segment = std::make_shared<Nfa>(segments[0]);
+        segment->trim();
+        if (segment->get_num_of_states() > 0 || include_empty) {
+            return {{ segment }};
+        } else {
+            return {};
         }
+    }
+
+    State unused_state = aut.get_num_of_states(); // get some State not used in aut
+
+    // segments_one_initial_final[init, final] is the pointer to automaton created from one of
+    // the segments such that init and final are one of the initial and final states of the segment
+    // and the created automaton takes this segment, sets initialstates={init}, finalstates={final}
+    // and trims it; also segments_one_initial_final[unused_state, final] is used for the first
+    // segment (where we always want all initial states, only final state changes) and
+    // segments_one_initial_final[init, unused_state] is similarly for the last segment
+    // TODO: should we use unordered_map? then we need hash
```

Comment:
unordered_map has better performance for certain function (as fair as I know), however, if there are not issues with this particular usage, you could keep the map.

Reply:
You are right. We use `std::unordered_map` in our refactorization of the noodlification. However, there are some issues I have to resolve first before the refactorization of noodlification can be merged into `devel` as well.

Reply:
This will come in a next PR. Marking as resolved for now.